proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -368,174 +368,139 @@ lemma invoke_sched_context_corres:
-   apply (clarsimp simp: objBits_simps)
-  apply (simp only: bind_assoc)
-  apply (subst bind_assoc_group4)
-  apply (rule monadic_rewrite_rewrite_head)
-   apply (rule monadic_rewrite_imp)
-    apply (rule monadic_rewrite_sym)
-    apply (rule getSchedContext_setSchedContext_decompose_decompose_ext2[unfolded K_bind_def])
-   apply (clarsimp simp: objBits_simps)
-  apply (simp only: bind_assoc)
-  apply (clarsimp simp: setRefillHd_def updateRefillHd_def bind_assoc)
-  apply (rule monadic_rewrite_imp)
-   apply (rule monadic_rewrite_trans_dup)
-    apply (rule monadic_rewrite_sym)
-    apply (rule getSchedContext_setSchedContext_decompose[unfolded K_bind_def])
-   apply (rule monadic_rewrite_refl3)
-   apply clarsimp
-  apply (fastforce simp: objBits_simps length_replaceAt)
-  done
-
 lemma getCurTime_sp:
   "\<lbrace>P\<rbrace> getCurTime \<lbrace>\<lambda>rv. P and (\<lambda>s. rv = ksCurTime s)\<rbrace>"
   by (wpsimp simp: getCurTime_def)
 
 lemma isRoundRobin_corres:
-  "corres (=) (pspace_aligned and pspace_distinct and sc_at sc_ptr) \<top>
+  "corres (=) (sc_at sc_ptr) (sc_at' sc_ptr)
               (is_round_robin sc_ptr) (isRoundRobin sc_ptr)"
-  apply (rule corres_cross[where Q' = "sc_at' sc_ptr", OF sc_at'_cross_rel], fastforce)
   apply (clarsimp simp: is_round_robin_def isRoundRobin_def)
   apply (corressimp corres: get_sc_corres
                       simp: sc_relation_def)
   done
 
+lemma valid_sc_size_sc_relation:
+  "\<lbrakk>valid_sched_context_size n; sc_relation sc n sc'\<rbrakk> \<Longrightarrow> n = objBits sc' - minSchedContextBits"
+  by (clarsimp simp: sc_relation_def objBits_simps valid_sched_context_size_def scBits_simps)
+
+(* updateAt *)
+
+lemma updateAt_index:
+  "\<lbrakk>xs \<noteq> []; i < length xs; j < length xs\<rbrakk>
+   \<Longrightarrow> (updateAt i xs f) ! j = (if i = j then f (xs ! i) else (xs ! j))"
+  by (fastforce simp: updateAt_def null_def nth_append)
+
+lemma wrap_slice_updateAt_eq:
+  "\<lbrakk>if start + count \<le> mx
+       then (i < start \<or> start + count \<le> i)
+       else (start + count - mx \<le> i \<and> i < start);
+    count \<le> mx; start < mx; mx \<le> length xs; xs \<noteq> []; i < mx\<rbrakk>
+   \<Longrightarrow> wrap_slice start count mx xs = wrap_slice start count mx (updateAt i xs new)"
+  apply (rule nth_equalityI)
+   apply clarsimp
+  by (subst wrap_slice_index; clarsimp simp: updateAt_index split: if_split_asm)+
+
+lemma take_updateAt_eq[simp]:
+  "n \<le> i \<Longrightarrow> take n (updateAt i ls f) = take n ls"
+  by (clarsimp simp: updateAt_def)
+
+lemma valid_obj'_scPeriod_update[simp]:
+  "valid_obj' (KOSchedContext (scPeriod_update (\<lambda>_. period) sc')) = valid_obj' (KOSchedContext sc')"
+  by (fastforce simp: valid_obj'_def valid_sched_context'_def valid_sched_context_size'_def objBits_simps)
+
+(* projection *)
+
+(* it would be good to rewrite all getting wp rules in this form *)
+lemma getSchedContext_wp':
```

Comment:
Maybe. The standard style is to combine the return value `rv` with `ko_at’`, via forall. Not quite sure if one is always better than the other.

Reply:
I'm not really sure either. `the` is dangerous, because it can escape the `sc_at'` and you can then go hunting for a long time for the thing that justified using `the` in the first place. The quantifier in `ko_at'` binds it to the justification and is usually nicer once you are within a pure HOL formula. But of course quantifiers are annoying in `wp`.

Reply:
I suppose it depends and optional. What I see quite often in Refine are situations where `sc_at’` is already available and, `!ko. ko_at` ko p s --> P ko` in postcondition is annoying in some wp proof for a setter, and it is possible to work around it by rewriting the postcondition to use a projection in place of `ko`. A getter wp rule in the projection style like here will let us skip such rewriting steps.

Reply:
We could probably make a lifting rule that takes a `ko_at'` precondition and turns it into a `typ_at'` + `the` precondition. This could then be used manually when we know it's safe.

(Not necessarily for this PR, though, happy for this rule to get merged separately)

Reply:
So, that is preferring `koTypeOf` over `projectKO_opt`? Both require the alignment conditions and I suppose that is necessary anyway.

Reply:
Either one could work as lifting rule. IIRC `obj_at' True` and `typ_at'` should be equivalent for the right type.

Reply:
ok, I see, it can be `sc_at’` as what we have here. So, what you are suggesting is to write a lifting rule that essentially converts the existing wp rules with `ko_at’` into the projection style here, is this correct? That would be nicely generic and useful.

Reply:
Yes, that's what I meant. I'm not entirely sure wether it's possible to do one generic lifting rule, because it's hard to generically talk about types, i.e. it's possible that you have simplify away preconditions once it is instantiated. But it'd be worth looking into at some point.

Reply:
The two preconditions are equivalent:
```
lemma forall_ko_at'_equiv_obj_at'_projection:
  "(λs. ∀ko::'a::pspace_storable. ko_at' ko p s ⟶ P ko s) =
        (λs. obj_at' (λ_::'a::pspace_storable. True) p s ⟶ P (the ((ksPSpace s |> projectKO_opt) p)) s)"
  by (fastforce simp: obj_at'_def projectKOs opt_map_red)
```
Probably it is quite ok to go back and forth between the two versions. Simply instantiating `'a` with `sched_context` will give `sc_at'` and `scs_of'`, for instance.